File: packages/adapters/src/codex-bridge.py
import json, os, pathlib, pwd, socket, subprocess, threading, time

root = pathlib.Path("/tmp/r2cloud-control")
root.mkdir(mode=0o700, exist_ok=True)
(root / "in").mkdir(exist_ok=True)
(root / "out").mkdir(exist_ok=True)
(root / "events").mkdir(exist_ok=True)
event_seq = 0
event_batch = []
turn_started = None
clients = set()
lock = threading.RLock()


def emit(value):
    data = (json.dumps(value) + "\n").encode()
    for client in list(clients):
        try:
            client.sendall(data)
        except OSError:
            clients.discard(client)
            client.close()


agent = pwd.getpwnam("r2-agent")
home = pathlib.Path(agent.pw_dir) / ".codex"
home.mkdir(mode=0o700, exist_ok=True)
os.chown(home, agent.pw_uid, agent.pw_gid)


def unprivileged():
    os.setgroups([])
    os.setgid(agent.pw_gid)
    os.setuid(agent.pw_uid)


proc = subprocess.Popen(
    [
        "codex",
        "app-server",
        "--listen",
        "stdio://",
        "-c",
        'cli_auth_credentials_store="file"',
    ],
    stdin=subprocess.PIPE,
    stdout=subprocess.PIPE,
    stderr=subprocess.DEVNULL,
    cwd=agent.pw_dir,
    preexec_fn=unprivileged,
    env={
        "PATH": "/opt/r2cloud/bin:" + os.environ["PATH"],
        "HOME": str(home),
        "CODEX_HOME": str(home),
        "LANG": "C.UTF-8",
    },
    text=True,
)


def save(name, value):
    temp = root / (name + ".tmp")
    temp.write_text(json.dumps(value))
    temp.replace(root / name)


def listen():
    global event_seq, event_batch
    for line in proc.stdout:
        if len(line) > 1048576:
            proc.kill()
            return
        try:
            message = json.loads(line)
            ident = message.get("id")
            if message.get("method"):
                event_seq += 1
                if event_seq > 20000:
                    proc.kill()
                    return
                if (event_seq - 1) % 10 == 0:
                    event_batch = []
                event_batch.append(
                    {
                        "seq": event_seq,
                        "message": message,
                        "providerElapsedMs": (
                            round((time.monotonic() - turn_started) * 1000)
                            if turn_started is not None
                            else None
                        ),
                    }
                )
                save(
                    "events/batch-" + str((event_seq - 1) // 10) + ".json", event_batch
                )
                save("events/head.json", {"seq": event_seq})
                with lock:
                    emit({"event": event_batch[-1]})
            if (
                isinstance(ident, str)
                and len(ident) == 64
                and (not message.get("method"))
            ):
                save("out/" + ident + ".json", message)
                with lock:
                    emit({"key": ident, "response": message})
            elif message.get("method") == "turn/completed":
                save("turn.json", message.get("params", {}))
            elif (
                message.get("method") == "item/completed"
                and message.get("params", {}).get("item", {}).get("type")
                == "agentMessage"
            ):
                save(
                    "message.json",
                    {"text": str(message["params"]["item"].get("text", ""))[-16000:]},
                )
            elif message.get("method") == "item/agentMessage/delta":
                save(
                    "progress.json",
                    {"text": str(message.get("params", {}).get("delta", ""))[-16000:]},
                )
        except (ValueError, OSError):
            proc.kill()
            return


threading.Thread(target=listen, daemon=True).start()
seen = set()


def lines(client):
    buffer = b""
    while True:
        try:
            data = client.recv(65536)
        except socket.timeout:
            continue
        if not data:
            return
        buffer += data
        if len(buffer) > 1048576:
            return
        while b"\n" in buffer:
            line, buffer = buffer.split(b"\n", 1)
            yield line


def accept(client):
    global turn_started
    client.settimeout(5)
    try:
        for line in lines(client):
            if len(line) > 1048576:
                break
            command = json.loads(line)
            with lock:
                if "cursor" in command:
                    cursor = int(command["cursor"])
                    for batch in range(cursor // 10, (event_seq + 9) // 10):
                        for entry in json.loads(
                            (
                                root / ("events/batch-" + str(batch) + ".json")
                            ).read_text()
                        ):
                            if entry["seq"] > cursor:
                                client.sendall(
                                    (json.dumps({"event": entry}) + "\n").encode()
                                )
                    clients.add(client)
                    client.sendall(b'{"ready":true}\n')
                    continue
                key = command["key"]
                if len(key) != 64 or any((c not in "0123456789abcdef" for c in key)):
                    break
                response = root / ("out/" + key + ".json")
                if response.exists():
                    client.sendall(
                        (
                            json.dumps(
                                {
                                    "key": key,
                                    "response": json.loads(response.read_text()),
                                }
                            )
                            + "\n"
                        ).encode()
                    )
                    continue
                if key in seen:
                    continue
                if command.get("lookup"):
                    client.sendall(
                        (json.dumps({"key": key, "unresolved": True}) + "\n").encode()
                    )
                    continue
                message = command["message"]
                seen.add(key)
                save("in/" + key + ".json", message)
                if message.get("method") == "turn/start":
                    turn_started = time.monotonic()
                proc.stdin.write(json.dumps(message) + "\n")
                proc.stdin.flush()
                if "id" not in message or "method" not in message:
                    ack = {"notified": True}
                    save("out/" + key + ".json", ack)
                    client.sendall(
                        (json.dumps({"key": key, "response": ack}) + "\n").encode()
                    )
    except (ValueError, OSError, KeyError):
        pass
    finally:
        with lock:
            clients.discard(client)
        client.close()


server = socket.socket(socket.AF_UNIX)
server.bind(str(root / "bridge.sock"))
server.listen(2)


def connections():
    while proc.poll() is None:
        client, _ = server.accept()
        threading.Thread(target=accept, args=(client,), daemon=True).start()


threading.Thread(target=connections, daemon=True).start()
proc.wait()
with lock:
    emit({"exit": proc.returncode})
save("exit.json", {"code": proc.returncode})

File: packages/adapters/src/codex-bridge.ts
import { chownSync, existsSync, mkdirSync, readFileSync, renameSync, writeFileSync } from 'node:fs';
import { spawn, execFileSync } from 'node:child_process';
import { createServer, type Socket } from 'node:net';
import type { Readable } from 'node:stream';

type Message = { id?: string | number; method?: string; params?: Record<string, any> };
type Event = { seq: number; message: Message; providerElapsedMs: number | null };
const root = '/tmp/r2cloud-control';
mkdirSync(root, { mode: 0o700, recursive: true });
for (const directory of ['in', 'out', 'events'])
  mkdirSync(`${root}/${directory}`, { recursive: true });
const [, , uid, gid, , directory] = execFileSync('getent', ['passwd', 'r2-agent'], {
  encoding: 'utf8',
})
  .trim()
  .split(':');
if (!uid || !gid || !directory) throw new Error('Agent user is unavailable.');
const home = `${directory}/.codex`;
mkdirSync(home, { mode: 0o700, recursive: true });
chownSync(home, Number(uid), Number(gid));
const agent = spawn(
  'setpriv',
  [
    '--reuid',
    uid,
    '--regid',
    gid,
    '--clear-groups',
    'codex',
    'app-server',
    '--listen',
    'stdio://',
    '-c',
    'cli_auth_credentials_store="file"',
  ],
  {
    cwd: directory,
    env: {
      PATH: `/opt/r2cloud/bin:${process.env.PATH}`,
      HOME: home,
      CODEX_HOME: home,
      LANG: 'C.UTF-8',
    },
    stdio: ['pipe', 'pipe', 'ignore'],
  },
);
const clients = new Set<Socket>();
const seen = new Set<string>();
let sequence = 0;
let batch: Event[] = [];
let turnStarted: number | undefined;
function save(path: string, value: unknown) {
  writeFileSync(`${root}/${path}.tmp`, JSON.stringify(value), { mode: 0o600 });
  renameSync(`${root}/${path}.tmp`, `${root}/${path}`);
}
function send(client: Socket, value: unknown) {
  if (client.destroyed) return;
  if (client.writableLength > 32 * 1024 * 1024) {
    client.destroy();
    return;
  }
  client.write(JSON.stringify(value) + '\n');
}
function emit(value: unknown) {
  for (const client of clients) send(client, value);
}
function lines(stream: Readable, accept: (value: any) => void, fail: () => void) {
  let buffer = '';
  stream.setEncoding('utf8');
  stream.on('data', (chunk: string) => {
    buffer += chunk;
    if (Buffer.byteLength(buffer) > 2 * 1024 * 1024) {
      fail();
      return;
    }
    let end: number;
    while ((end = buffer.indexOf('\n')) !== -1) {
      const line = buffer.slice(0, end);
      buffer = buffer.slice(end + 1);
      try {
        if (Buffer.byteLength(line) > 1024 * 1024)
          throw new Error('Bridge message exceeds its limit.');
        accept(JSON.parse(line));
      } catch {
        fail();
        return;
      }
    }
  });
  stream.on('error', fail);
}
lines(
  agent.stdout,
  (message: Message) => {
    if (message.method) {
      if (++sequence > 20000) throw new Error('Provider event limit reached.');
      if ((sequence - 1) % 10 === 0) batch = [];
      const event = {
        seq: sequence,
        message,
        providerElapsedMs:
          turnStarted === undefined ? null : Math.round(performance.now() - turnStarted),
      };
      batch.push(event);
      save(`events/batch-${Math.floor((sequence - 1) / 10)}.json`, batch);
      save('events/head.json', { seq: sequence });
      emit({ event });
    }
    const params = message.params ?? {};
    if (typeof message.id === 'string' && /^[a-f0-9]{64}$/.test(message.id) && !message.method) {
      save(`out/${message.id}.json`, message);
      emit({ key: message.id, response: message });
    } else if (message.method === 'turn/completed') save('turn.json', params);
    else if (message.method === 'item/completed' && params.item?.type === 'agentMessage')
      save('message.json', { text: String(params.item.text ?? '').slice(-16000) });
    else if (message.method === 'item/agentMessage/delta')
      save('progress.json', { text: String(params.delta ?? '').slice(-16000) });
  },
  () => agent.kill('SIGKILL'),
);
const server = createServer((client) => {
  client.on('close', () => clients.delete(client));
  lines(
    client,
    (command) => {
      if ('cursor' in command) {
        const cursor = command.cursor;
        if (!Number.isSafeInteger(cursor) || cursor < 0 || cursor > sequence)
          throw new Error('Invalid replay cursor.');
        for (let index = Math.floor(cursor / 10); index < Math.ceil(sequence / 10); index++) {
          const events: Event[] = JSON.parse(
            readFileSync(`${root}/events/batch-${index}.json`, 'utf8'),
          );
          for (const event of events) if (event.seq > cursor) send(client, { event });
        }
        clients.add(client);
        send(client, { ready: true });
        return;
      }
      const { key, message } = command;
      if (typeof key !== 'string' || !/^[a-f0-9]{64}$/.test(key))
        throw new Error('Invalid command key.');
      const path = `${root}/out/${key}.json`;
      if (existsSync(path)) {
        send(client, { key, response: JSON.parse(readFileSync(path, 'utf8')) });
        return;
      }
      if (seen.has(key)) return;
      if (command.lookup) {
        send(client, { key, unresolved: true });
        return;
      }
      if (!message || typeof message !== 'object') throw new Error('Invalid command.');
      seen.add(key);
      save(`in/${key}.json`, message);
      if (message.method === 'turn/start') turnStarted = performance.now();
      if (agent.stdin.writableLength > 1024 * 1024) throw new Error('Provider input is blocked.');
      agent.stdin.write(JSON.stringify(message) + '\n');
      if (message.id === undefined || !message.method) {
        const response = { notified: true };
        save(`out/${key}.json`, response);
        send(client, { key, response });
      }
    },
    () => client.destroy(),
  );
});
server.listen(`${root}/bridge.sock`);
function finish(code: number | null) {
  save('exit.json', { code });
  emit({ exit: code ?? 1 });
  for (const client of clients) client.end();
  server.close();
}
agent.on('error', () => agent.kill('SIGKILL'));
agent.stdin.on('error', () => agent.kill('SIGKILL'));
agent.on('close', finish);
server.on('error', () => agent.kill('SIGKILL'));
